modules/HapticGlove_module/src/KalmanFilter.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file KalmanFilter.cpp
+ * @authors  Kourosh Darvish <kourosh.darvish@iit.it>
+ * @copyright 2020 iCub Facility - Istituto Italiano di Tecnologia
+ *            Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ * @date 2020
+ */
+
+#include <KalmanFilter.hpp>
+#include <iostream>
+using namespace HapticGlove;
+KalmanFilter::KalmanFilter(const double dt,
+                           const size_t n,
+                           const size_t p,
+                           const size_t m,
+                           const Eigen::MatrixXd& F,
+                           const Eigen::MatrixXd& G,
+                           const Eigen::MatrixXd& H,
+                           const Eigen::MatrixXd& R,
+                           const Eigen::MatrixXd& Q)
+    : m_dt(dt)
+    , m_n(n)
+    , m_p(p)
+    , m_m(m)
+    , m_F(F)
+    , m_G(G)
+    , m_H(H)
+    , m_R(R)
+    , m_Q(Q)
+{
+    m_logPrefix = "KalmanFilter:: ";
+
+    m_Phi = Eigen::MatrixXd::Identity(n, n) + m_F * m_dt;
+    m_Gamma = m_G * m_dt;
+    m_Ht_Rinv = m_H.transpose() * m_R.inverse();
+    m_Ht_Rinv_H = m_H.transpose() * m_R.inverse() * m_H;
+    Gamma_Q_GammaT = m_Gamma * m_Q * m_Gamma.transpose();
+
+    m_w_bar = Eigen::MatrixXd::Zero(m_m, 1);
+}
+
+KalmanFilter::~KalmanFilter()
+{
+}
+
+bool KalmanFilter::initialize(const Eigen::MatrixXd& x0, const Eigen::MatrixXd& M0)
+{
+    m_x_bar = x0;
+    m_M = M0;
+
+    m_P = (m_M.inverse() + m_Ht_Rinv_H).inverse();
+    m_K = m_P * m_Ht_Rinv;
+    m_M = m_Phi * m_P * m_Phi.transpose() + m_Gamma * m_Q * m_Gamma.transpose();
+
+    return true;
+}
+
+bool KalmanFilter::setNewMeasurements(const Eigen::MatrixXd& z)
+{
+    m_z = z;
+    return true;
+}
+
+bool KalmanFilter::estimateNextState(const Eigen::MatrixXd& z, Eigen::MatrixXd& x_hat)
+{
+    /*
+     * J= 1/2 [(x-x_bar) M^(-1)(x-x_bar) + (z-Hx) R^(-1)(z-Hx)]
+     */
+
+    m_z = z;
+
+    m_P = (m_M.inverse() + m_Ht_Rinv_H).inverse();
+    m_K = m_P * m_Ht_Rinv;
+    m_x_hat = m_x_bar + m_K * (m_z - m_H * m_x_bar);
+    m_x_bar = m_Phi * m_x_hat + m_Gamma * m_w_bar;
+    m_M = m_Phi * m_P * m_Phi.transpose() + m_Gamma * m_Q * m_Gamma.transpose();
+
+    x_hat = m_x_hat;
+
+    return true;
+}
+
+bool KalmanFilter::estimateNextState(const Eigen::MatrixXd& z)
+{
+    /*
+     * J= 1/2 [(x-x_bar) M^(-1)(x-x_bar) + (z-Hx) R^(-1)(z-Hx)]
+     */
+
+    m_z = z;
+
+    m_P = (m_M.inverse() + m_Ht_Rinv_H).inverse();
+    m_K = m_P * m_Ht_Rinv;
+    m_x_hat = m_x_bar + m_K * (m_z - m_H * m_x_bar);
+    m_x_bar = m_Phi * m_x_hat + m_Gamma * m_w_bar;
+    m_M = m_Phi * m_P * m_Phi.transpose() + m_Gamma * m_Q * m_Gamma.transpose();
```

Comment:
Note, if these computations are called during the control loop, Eigen might need to create temporary objects that may slow down the execution (because they need dynamic memory allocation). See https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html

I am not sure this is the case anyways

Reply:
Yes, this is called in the control loop. I have added `noalias()` method before the `=` operator, to take care of that when having a matrix product or inversion. I have made this update all around code.
Thanks for pointing this out.

Reply:
You can use what defined in http://eigen.tuxfamily.org/index.php?title=FAQ#Where_in_my_program_are_temporary_objects_created.3F to understand if some temporary objects are created. If there are, the suggestion would be to use variables to store the intermediate steps.

Reply:
@S-Dafarra  @traversaro   I added the assertion check as in http://eigen.tuxfamily.org/index.php?title=FAQ#Where_in_my_program_are_temporary_objects_created.3F . However, I did the check with or WITHOUT `.noalias()` ; both the times the results were similar. No message showing the heap memory allocation. However, for the matrix product it should be case. Am I doing sth wrong?
Here is the commit  https://github.com/robotology/walking-teleoperation/pull/72/commits/ab05a12f3c8096c9810aff5e2f6fc903e855b5d1 
 
I added permanently the option to CMake and made it as false by default.

I had some compilation error while adding the command as mentioned in  https://github.com/stack-of-tasks/tsid/issues/7 and benefit from the implementation in https://github.com/stack-of-tasks/tsid/blob/b0d6bff80292fb3451ca7ca438a4ab84b5b8c022/CMakeLists.txt and https://github.com/stack-of-tasks/tsid/blob/devel/include/tsid/math/fwd.hpp

Reply:
To be honest I never tried it myself. From my side, this is not blocking at all.

IN any case, have you tried compiling in debug? Or intentionally add some other matrix in the product?

Reply:
No, I tried in Release mode, not in `Debug` mode. 

> Or intentionally add some other matrix in the product?

No, I just remove `.noalias()` from the member variables.

Reply:
OK, I tried in debug mode and returns the assertion results. The assertion fails in the following line:
```
    m_P = (m_M.inverse() + m_Ht_Rinv_H).inverse();
```
which is expected.

The following line the assertion passes:
```
    m_K = m_P * m_Ht_Rinv;
```
While again, the following line the assertion does not pass, which would expect it to pass:
```
    m_x_hat = m_x_bar + m_K * (m_z - m_H * m_x_bar);
```
and here is the back-trace results:

```
HapticGloveModule: /usr/include/eigen3/Eigen/src/Core/util/Memory.h:143: void Eigen::internal::check_that_malloc_is_allowed(): Assertion `is_malloc_allowed() && "heap allocation is forbidden (EIGEN_RUNTIME_NO_MALLOC is defined and g_is_malloc_allowed is false)"' failed.

Thread 1 "HapticGloveModu" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff75d1859 in __GI_abort () at abort.c:79
#2  0x00007ffff75d1729 in __assert_fail_base (fmt=0x7ffff7767588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=0x555555627490 "is_malloc_allowed() && \"heap allocation is forbidden (EIGEN_RUNTIME_NO_MALLOC is defined and g_is_malloc_allowed is false)\"", 
    file=0x555555627458 "/usr/include/eigen3/Eigen/src/Core/util/Memory.h", line=143, function=<optimized out>) at assert.c:92
#3  0x00007ffff75e2f36 in __GI___assert_fail (
    assertion=0x555555627490 "is_malloc_allowed() && \"heap allocation is forbidden (EIGEN_RUNTIME_NO_MALLOC is defined and g_is_malloc_allowed is false)\"", 
    file=0x555555627458 "/usr/include/eigen3/Eigen/src/Core/util/Memory.h", line=143, function=0x555555627420 "void Eigen::internal::check_that_malloc_is_allowed()")
    at assert.c:101
#4  0x000055555557512b in Eigen::internal::check_that_malloc_is_allowed () at /usr/include/eigen3/Eigen/src/Core/util/Memory.h:143
#5  0x0000555555575144 in Eigen::internal::aligned_malloc (size=24) at /usr/include/eigen3/Eigen/src/Core/util/Memory.h:155
#6  0x0000555555580105 in Eigen::internal::conditional_aligned_malloc<true> (size=24) at /usr/include/eigen3/Eigen/src/Core/util/Memory.h:214
#7  0x000055555557e00d in Eigen::internal::conditional_aligned_new_auto<double, true> (size=3) at /usr/include/eigen3/Eigen/src/Core/util/Memory.h:374
#8  0x000055555557a398 in Eigen::DenseStorage<double, -1, -1, -1, 1>::resize (this=0x55555587e968, size=3, rows=3, cols=1)
    at /usr/include/eigen3/Eigen/src/Core/DenseStorage.h:408
#9  0x0000555555577356 in Eigen::PlainObjectBase<Eigen::Matrix<double, -1, -1, 1, -1, -1> >::resize (this=0x55555587e968, rows=3, cols=1)
    at /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:293
#10 0x0000555555589050 in Eigen::internal::resize_if_allowed<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, double, double> (
    dst=..., src=...) at /usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:720
#11 0x0000555555585412 in Eigen::internal::call_dense_assignment_loop<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::internal::assign_op<double, double> > (dst=..., src=..., func=...) at /usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:734
#12 0x00005555555837fa in Eigen::internal::Assignment<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::internal::assign_op<double, double>, Eigen::internal::Dense2Dense, void>::run (dst=..., src=..., func=...) at /usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:879
#13 0x00005555555823f9 in Eigen::internal::call_assignment_no_alias<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::internal::assign_op<double, double> > (dst=..., src=..., func=...) at /usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:836
#14 0x00005555555d036b in Eigen::internal::assignment_from_xpr_op_product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_difference_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, 0> const>, 0>, Eigen::internal::assign_op<double, double>, Eigen::internal::add_assign_op<double, double> >::run<Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_difference_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, 0> const>, 0> const>, Eigen::internal::assign_op<double, double> > (dst=..., src=...) at /usr/include/eigen3/Eigen/src/Core/ProductEvaluators.h:223
#15 0x00005555555cf813 in Eigen::internal::call_assignment_no_alias<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_difference_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, 0> const>, 0> const>, Eigen::internal::assign_op<double, double> > (dst=..., src=..., func=...) at /usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:836
#16 0x00005555555ce608 in Eigen::NoAlias<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::MatrixBase>::operator=<Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_difference_op<double, double>, Eigen::Matrix<double, -1, -1, 1, -1, -1> const, Eigen::Product<Eigen::Matrix<double, -1, -1, 1, -1, -1>, Eigen::Matrix<double, -1, -1, 1, -1, -1>, 0> const>, 0> const> > (this=0x7fffffffc570, other=...) at /usr/include/eigen3/Eigen/src/Core/NoAlias.h:42
#17 0x00005555555cd0f3 in HapticGlove::KalmanFilter::estimateNextState (this=0x55555587e820, z=...)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/KalmanFilter.cpp:95
#18 0x00005555555de666 in HapticGlove::Estimator::estimateNextState (this=0x5555558435a0, z=...)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/MotorEstimation.cpp:82
#19 0x00005555555e0fba in HapticGlove::Estimators::estimateNextState (this=0x555555755720, z=std::vector of length 9, capacity 9 = {...})
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/RobotMotorsEstimation.cpp:97
#20 0x0000555555574372 in HapticGlove::RobotController::estimateNextStates (this=0x5555556e65f0)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/RobotController.cpp:511
#21 0x000055555561ee76 in HapticGlove::Teleoperation::getFeedbacks (this=0x555555712bc0)
--Type <RET> for more, q to quit, c to continue without paging--
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/Teleoperation.cpp:184
#22 0x000055555561f3ac in HapticGlove::Teleoperation::run (this=0x555555712bc0)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/Teleoperation.cpp:239
#23 0x00005555555c6b7e in HapticGloveModule::updateModule (this=0x7fffffffc910)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/HapticGloveModule.cpp:146
#24 0x00007ffff7c02dc4 in yarp::os::RFModule::runModule() () from /home/kourosh/icub_ws/robotology-superbuild/build/install/lib/libYARP_os.so.3
#25 0x00007ffff7c0328e in yarp::os::RFModule::runModule(yarp::os::ResourceFinder&) ()
   from /home/kourosh/icub_ws/robotology-superbuild/build/install/lib/libYARP_os.so.3
#26 0x000055555556dff8 in main (argc=1, argv=0x7fffffffcb48)
    at /home/kourosh/icub_ws/robotology-superbuild/src/walking-teleoperation/modules/HapticGlove_module/src/main.cpp:40
(gdb) 
```
here it does not  create temporary objects:
`    m_x_bar = m_Phi * m_x_hat + m_Gamma * m_w_bar;
`
while in `    m_M = m_Phi * m_P * m_Phi.transpose() + m_Gamma * m_Q * m_Gamma.transpose();` it does make new objects. I tried to introduce new variable for `m_Phi.transpose()` and `m_Gamma.transpose()` but for those parts also new objects are dynamically allocated. 
I stop the analysis at this point. I will keep the `noalias()` as it should not hurt  as it has been carefully used.

Reply:
So, finally I am going to keep the changes introduced in this commit https://github.com/robotology/walking-teleoperation/pull/72/commits/e94eb10ad4cbdf3b47ea37914ef41cbab0faa2cb , as it will not change anything if the option is not set `ON` by the user and if it is not compiled in `Debug` mode. I will not do further optimization at this stage, since it is not limiting. @S-Dafarra if you agree I resolve this conversation with the outputs of the investigation that has been done.

Reply:
Yes ok for me. 
Just notice that the fact that the error disappears when compiling in Release does not mean that it is fixed. When compiling in Release all the asserts (including the one above) disappear. So, the issue is still there, simply you "silenced" the message.

Approximately, the only operations that do not trigger a memory allocation are the sum, the transpose and a simple matrix product. The sum of two matrix products does (see https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html). 
This is why 
```
    m_x_hat = m_x_bar + m_K * (m_z - m_H * m_x_bar);
```
produces a temporary object. You need to store the result of ``m_H * m_x_bar`` somewhere before applying the difference. The same also for ``m_K * (m_z - m_H * m_x_bar)``.

Basically, it becomes a tradeoff between the readability of the code and its execution speed.

Reply:
> Just notice that the fact that the error disappears when compiling in Release does not mean that it is fixed. When compiling in Release all the asserts (including the one above) disappear. So, the issue is still there, simply you "silenced" the message.


Yes, I know!


> Approximately, the only operations that do not trigger a memory allocation are the sum, the transpose and a simple matrix product. The sum of two matrix products does (see https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html).
This is why
```
    m_x_hat = m_x_bar + m_K * (m_z - m_H * m_x_bar);
```
produces a temporary object. You need to store the result of m_H * m_x_bar somewhere before applying the difference. The same also for m_K * (m_z - m_H * m_x_bar).

I tried last night, it needs many intermediate variables to add. I am not in the favour of it as it is not a blocking point at this stage, as you mentioned, it will be readable! Probably, even this one `m_K * (m_z - m_H * m_x_bar).` needs an intermediate member variable, since we have two matrix product.

For example I was trying to do it for `m_Phi * m_P * m_Phi.transpose() + m_Gamma * m_Q * m_Gamma.transpose();` we need one variable for each transpose, and then another one for single matrix product  and then another one for each operand of `+` and ... . I really prefer in this way to be readable at this stage.